spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,367 @@
+  // an estop endpoint.
+  const auto estop_status = robot_->IsEstopped();
+  if (!estop_status) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not check estop status");
+    return false;
+  }
+  if (estop_status.response) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Robot is e-stopped, cannot continue.");
+    return false;
+  }
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Estop check complete!");
+  return true;
+}
+
+bool SpotHardware::get_lease() {
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Getting Lease");
+  // First create a lease client.
+  ::bosdyn::client::Result<::bosdyn::client::LeaseClient*> lease_client_resp =
+      robot_->EnsureServiceClient<::bosdyn::client::LeaseClient>();
+  if (!lease_client_resp) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not create lease client");
+    return false;
+  }
+  lease_client_ = lease_client_resp.response;
+  // Then acquire the lease for the body.
+  const auto lease_res = lease_client_->AcquireLease("body");
+  if (!lease_res) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not acquire body lease");
+    return false;
+  }
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Lease acquired!!");
+  return true;
+}
+
+bool SpotHardware::power_on() {
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Powering on...");
+  const auto power_status = robot_->PowerOnMotors(std::chrono::seconds(60), 1.0);
+  if (!power_status) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not power on the robot");
+    return false;
+  }
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Powered on!");
+  return true;
+}
+
+void state_stream_loop(std::stop_token stop_token, ::bosdyn::client::RobotStateStreamingClient* stateStreamClient,
+                       StateHandler&& state_policy) {
+  ::bosdyn::api::RobotStateStreamResponse latest_state_stream_response;
+
+  while (!stop_token.stop_requested()) {
+    // Get robot state stream
+    auto robot_state_stream = stateStreamClient->GetRobotStateStream();
+    if (!robot_state_stream) {
+      RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Failed to get robot state");
+      continue;
+    }
+    latest_state_stream_response = std::move(robot_state_stream.response);
+    state_policy(latest_state_stream_response);
+  }
+}
+
+bool SpotHardware::start_state_stream(StateHandler&& state_policy) {
+  // Start state streaming
+  auto robot_state_stream_client_resp = robot_->EnsureServiceClient<::bosdyn::client::RobotStateStreamingClient>();
+  if (!robot_state_stream_client_resp) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not create robot state client");
+    return false;
+  }
+  state_client_ = robot_state_stream_client_resp.move();
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Robot State Client created");
+
+  state_thread_ = std::jthread(&spot_ros2_control::state_stream_loop, state_client_, state_policy);
```

Comment:
@khughes-bdai meta: what is taking care of synchronizing read/writes to the hardware interface states?

Reply:
To the best of my knowledge, I think this is handled by the `controller_manager` node? I know we can specify the update rate of the `joint_state_broadcaster` through the config yaml we are using, which I assume under the hood is going through the controller manager to request reads to the hardware interface. But I am not sure exactly how synchronization between command/state is handled.

Reply:
> To the best of my knowledge, I think this is handled by the controller_manager node?

Hmm, I don't think that's the case. The controller manager will read state and write commands sequentially, yes, but we have an `std::jthread` writing states. What if the controller manager reads state while the `std::jthread` is writing it? We'll read corrupted state.

Perhaps we can check how UMV deals with this? I'd naively suggest double buffering, but perhaps mutexes aren't that expensive.

Reply:
Judging from the spot rl code, it looks like we only need the separate thread for getting the robot state stream -- commands I think you can just send directly (https://github.com/bdaiinstitute/spot-rl/blob/cfee0ae675b2abbbc8d233322de391afe2036543/cpp/src/spot.cpp#L138). We can make a ticket for this to investigate some more @tcappellari-bdai

Reply:
FWIW I see UMV using mutexes, see the [`PlanarizerHardwareInterface`](https://github.com/bdaiinstitute/control_suite/blob/cec1673fa2295932d9bf6c7da60f7715efce2f94/src/cs_ros2_control/umv/djhotboing/hardware_interfaces/planarizer_hardware_interface/src/planarizer_hardware_interface.cpp#L171-L172) and the [`planarizer_encoder_bridge`](https://github.com/bdaiinstitute/control_suite/blob/cec1673fa2295932d9bf6c7da60f7715efce2f94/src/cs_ros2_control/umv/djhotboing/hardware_interfaces/planarizer_common/src/encoder_reader_lib.cpp#L316-L319) implementations.

Reply:
UMV uses mutexes. I'd strongly suggest we use one for the `StateStreamingHandler`.